keps/1337-reclaim-within-cohort-while-borrowing/README.md

```diff
@@ -0,0 +1,482 @@
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+## Motivation
+
+Currently, there is no support for preemption to reclaim quota within cohort
+for workloads which require borrowing. This is too constraining for some setups.
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- support preemption to reclaim quota within cohort for workloads that require borrowing.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- allow preemption of workloads, from other ClusterQueues, which are not borrowing.
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Introduce a new ClusterLevel API to support preemption while borrowing. The
+API proposes additional configuration to the `.preemption.reclaimWithinCohort`
+option.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As an administrator of Kueue in my organization I manage the following setup
+with two teams, `A` and `B`. There is a single cohort, and both teams have a
+pair of ClusterQueues in the cohort, called `Standard` and `Best-Effort`. There
+is also one ClusterQueue, called `Shared`, which is shared between the teams.
+
+All the `Standard` and `Best-Effort` ClusterQueues have `nominalQuota=0`, and
+`borrowingLimit>0`. The actual quota is exposed by the `Shared` ClusterQueue.
+The `Shared` ClusterQueue has `nominalQuota>0`, and `borrowingLimit=0`. This
+setup is depicted by the table:
```

Comment:
For the setup you described we don't need to have a Standard and Best-Effort CQs, one CQ with appropriate threshold is enough, correct?

Reply:
It would not be enough to justify `MaxPriorityThreshold`. If we had just two CQs, one with high priorities, one with low priorities, then `LowerPriority` would be enough.  Since we would only have one `Standard` CQ the preemption within would be controlled by `preemption.withinClusterQueue`, rather than `borrowWithinCohort`.

So, it is important in this setup, that we have two CQs with high priority workloads (`Standard`). Now, if we only had `LowerPriority`, then it would still allow that a workload in `A_Standard` preempts a workload in `B_Standard`. By introducing `MaxPriorityThreashold` between `Standard` and `Best-Effort` we make sure no workloads are preempted in `Standard` CQs.

Reply:
I see, we don't allow preemption within the CQ here